crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2717,6 +2717,21 @@ impl<'db> TypeInferenceBuilder<'db> {
         // - `[ast::CompOp::Is]`: return `false` if unequal, `bool` if equal
         // - `[ast::CompOp::IsNot]`: return `true` if unequal, `bool` if equal
         match (left, right) {
+            (Type::Union(union), other) => Some(UnionType::from_elements(
+                self.db,
+                union
+                    .elements(self.db)
+                    .iter()
+                    .filter_map(|element| self.infer_binary_type_comparison(*element, op, other)),
```

Comment:
A `None` return from `infer_binary_type_comparison` means "op not supported for these operands." If `infer_compare_expression` gets `None` it will emit a diagnostic and fall back to `Unknown` (if there was a comparison error we can't really say what the result should be.)

Using `filter_map` here silently excludes the `None` results from consideration, meaning we won't emit a diagnostic when we should, and we will provide an "answer" when we should fall back to `Unknown`.

I think instead we should short-circuit and return `None` if we see any `None`.

Probably we also want to change this at some point to return a `Result` instead of `Option`, both simply because it's more explicit, and because then the `Err` variant can propagate details of precisely which two "inner" types we failed to compare, in the recursive cases, allowing us to offer a better diagnostic.